### ISSUE TITLE
Running R files as scripts: warn about workspaces

### ISSUE THREAD

**Issue by mathew-hall** [rcgsheffield/sheffield_hpc#706]
R's default behaviour for workspace saving/loading might cause reproducibility problems. The `-no-save` and `--no-restore` options turn it off for `Rscript` and `R CMD BATCH`.

**Comment by willfurnass** [comment]
@mathew-hall A good suggestion!

**Comment by annakrystalli** [comment]
Thanks for the PR @mathew-hall! @willfurnass when you're back, should we have a look at whether we can turn this behaviour off globally? I can figure the **R** part if you can help with the **global** part?

**Comment by mikecroucher** [comment]
This is an important piece of documentation (thanks @mathew-hall ) but I am against changing the global behaviour.   Two reasons:

* You should never change the default behaviour of an installed application that may be in use.  Imagine if you were running a parameter sweep that relied on that behaviour and IT support changed the rules of the game mid-run.  Not Cool! Ever!

* It's the default behaviour! Isn't this what users of R will be expecting to happen?

Thoughts?

**Comment by annakrystalli** [comment]
In interactive mode, perhaps, but even then it's advised against for reproducibility reasons. In Rstudio atleast you have the option to say you don't want it saved on shut out of a session. On the cluster, especially for a batch job it doesn't seem to make sense to save the workspace in my view. I don't feel it is something folks that submit batch jobs will expect either.

**Comment by annakrystalli** [comment]
But that's just my feeling as a user. If there are sys admin reasons to avoid this then I defer to your experience.

**Comment by mikecroucher** [comment]
We could change it for future installs of R.  Let's see what Will and Matt think here too.

**Comment by annakrystalli** [comment]
👍

**Comment by willfurnass** [comment]
Think the way enable this behaviour for all R installs would be to create a `${R_HOME}/etc/.Rprofile.site file` containing:
```R
q <- function (save = "no", status = 0, runLast = TRUE)
.Internal(quit(save, status, runLast))
```
where $R_HOME is the prefix R is installed into e.g. `/usr/local/packages/apps/R/3.4.0/intel-17.0-sequential`.  

*However* if we implemented this then the difference in behaviour between R on HPC and local R might confuse people.  I think here it makes more sense to educate people on how to establish more idempotent workflows by manually specifying `--no-save --no-restore` (or `--vanilla`?) though
 * short blog post on the RSE site
 * mail to hpc users mailing list making them aware of this post.

Oh, and thanks @mathew-hall for the PR

**Comment by mathew-hall** [comment]
So, I think deferring to Linus on not breaking userland makes sense. That said, it's possible for there to be a difference in behaviour between HPC R and an R on someone's machine with the feature left enabled. For example, a user runs a job on HPC R that writes to variable `foo` (saved in a workspace). They then modify the script so the value `foo` is never assigned, but is still read from. If the script's run in the working directory it won't fail, but will fail anywhere else. Arguably the script is broken at the point it depends on a workspace image, so having it fail sooner might be preferable?

I think education is probably the best solution, maybe supported with a warning instead of changing behaviour? It's definitely best practice to turn the feature off, but I'm not sure if doing it on the user's behalf is guaranteed to be benign in all cases. Especially for batch jobs where there'd be no advance warning. 

If there isn't one already it might make sense to define an HPC policy on turning off dangerous features to protect the average user but letting people aware of the caveat opt-in (c.f. [`hdparm`'s `yes-i-know-what-i-am-doing`](https://www.thomas-krenn.com/en/wiki/SSD_Over-provisioning_using_hdparm) switch)?

tl;dr: I think @willfurnass has it right, it depends on how it's communicated to users, but education is probably the way to go.

**Comment by drj11** [comment]
I had always assumed that people writing scripts in R use `Rscript`, where the defaults are sensible.

If they don't, can we emphasise that they should?
